tools/libxl/libxl_arm.c

```diff
@@ -786,29 +803,84 @@ static int copy_partial_fdt(libxl__gc *gc, void *fdt, void *pfdt)
     return 0;
 }
 
-static int copy_coprocs_nodes(libxl__gc *gc, void *fdt, void *pfdt,
-                              const libxl_domain_build_info *info)
-{
-    int i, rc;
-
-    LOG(DEBUG, "Copy coprocs nodes from the partial FDT");
+struct backlink {
+    int nodeoff;
+    int copied;
+    struct backlink *parent;
+};
 
-    for (i = 0; i < info->num_coprocs; i++) {
-        const libxl_device_coproc *coproc = &info->coprocs[i];
+static int copy_ancestor(libxl__gc *gc, void *fdt, void *pfdt,
+                         struct backlink *nodelink)
+{
+    int ret = 0;
+
+    if (nodelink && nodelink->copied == 0)
+    {
+        copy_ancestor(gc, fdt, pfdt, nodelink->parent);
+        LOG(DEBUG, "Begin node %s", fdt_get_name(pfdt, nodelink->nodeoff,
+            NULL));
+        ret = fdt_begin_node(fdt, fdt_get_name(pfdt, nodelink->nodeoff, NULL));
```

Comment:
You call fdt_get_name(pfdt, nodelink->nodeoff, NULL) twice. Maybe introduce a variable?